### DISCUSSION TITLE
igl::harmonic and boundary conditions

### DISCUSSION THREAD

**Discussion by wtholliday** [libigl/libigl#2416] [Q&A] [answered]
I subdivided a tet a few times and ran `igl::harmonic` on it, thinking I would get C^2 continuity in my result, which should be more like a sphere (right?), but I got this with k=3.  I've just constrained the original 4 vertices of the tet. Using `igl::harmonic(F,B,B_values,3,W)`, so not passing initial vertex positions other than the 4 constrained vertices.

<img width="652" alt="image" src="https://github.com/user-attachments/assets/2863ee59-f7b9-49e3-8c87-463dd2c7bc98">

If I projected a few of the vertices around the original 4 onto a sphere and constrained those, perhaps I would end up with more of a sphere? The example has many more constrained vertices:

![image](https://github.com/user-attachments/assets/dcd173d9-8a41-4b34-bd17-f7dea3f34082)

More generally, how can I solve for a mesh that minimizes curvature (subject to constraints) using igl?

**Comment by alecjacobson** [comment, marked as answer]
This is a great question and the reason the solution looks non-smooth can be surprising at first.

The tutorial is smooth because we're defining the squared laplacian energy (minimized by biharmonic functions) over a smooth surface (the base surface is a plane so clearly smooth).

In your example of a tet, the base surface already has sharp edges and sharp kinks. The sharp edges are not so bad because they can be "developed" (or flattened) to a smooth surface (again the plane) locally without distortion. For this reason the biharmonic solutions are indeed smooth across the edges. But the points are trouble, here the surface that we're defining the laplacian upon (the original tet) is not smooth and not developable at the points. So a single biharmonic solve will not remove the sharpness. In a way, the sharpness is baked in.

If the goal is to smooth out sharp features entirely, then previous have tried to iterate on solving the biharmonic positions hoping that the sharp parts go away. Unfortunately this can be hard to wrangle.

Better, would be to measure more directly what you're trying to get rid of. This is how robust fairing methods like "Robust Fairing via Conformal Curvature Flow" or physics based shell/cloth energies work. Either of those could be implemented using libigl but don't have a direct function right now
